tests/postgresql-tests.py

```diff
@@ -44,6 +44,14 @@ def parse_options():
     parser.add_option('--azure', action="store_true", default=False,
                       help='If present, will connect as Azure, otherwise as '
                            'standard MSSQL')
+    parser.add_option('--debugging', action="store_true", default=False,
+                      help='If present, will display debug information and pause '
```

Comment:
Could we make the pause an option that is not enabled by default?

I would like to enable debugging builds at our https://jenkins.juliogonzalez.es/job/tds_fdw-build-free/, which basically allow us to run the whole set of tests (for all distributions and PostgreSQL option).

Reply:
In that cas, why won't we remove debugging param from run_tests and get rid of `if debugging` ?

Reply:
What do you mean? At CI we'd want debugging but not pause (so we get all data), and developers most likely would like debugging AND pause, so I think we need need both.

Reply:
Sorry, I was unclear @juliogonzalez : if CI needs details *any time* we could givind up the if statement and keep its content. This way, CI  will have details in case of test failure anytime. The only use of debugging param in the run_test function was for this purpose. But...
Looking closer, I have introduced a new bug here as this function is also called by mssql side and pymssql dien not have this extension. So I'll have to check exception type/class.

This said, --debugging option should be kept only for pid/pause purpose imho. And I have to check if I tell the user he/she has to press a key.

Anyway we could opt for your solution also.

Is it clearer ?

Sorry, i'm afk and use my phone to reply.

Reply:
Much clear now.

`tds_fdw-build` (job that runs after PR merges and daily) and `tds_fdw-build-pr` (jub that runs for each new PR, each new commit to a PR, or magic comment) are jobs to run without debugging, and without pauses.

`tds_fdw-build-free` is a job that is executed manually, and is available to common contributors (I could add you as well, if Geoff is fine). Benefit of that one, is that the developer can pick up what OS to test, what PostgreSQL to test, what git refs and repository to use for `ci-build` and `tds_fdw`... and now it should allow enabling debugging... but without the pauses. Even without the pauses, the info would be useful for an initial debugging (specially keeping in mind I am thinking about extracting `dmesg` and `postgresql.log` in such case.)

And finally, we have the developers, who must likely want both debugging AND pauses.

Long run, I'd also want to check if we can configure `tds_fdw-build-pr` to rerun a debugging build with the magic comment.

Reply:
Alright it's crystal clear for me too now. I did not figure all the CI things out! 
So I changed params :  

  * `--debugging` will have the purpose for developper to display backend pid and pause (and more maybe in the future ?)
  * `--details` will display detailled errors at test failure

Reply:
BTW my remark on mssql and the introduced bug is not  relevant  (I hadn't time to check the code with my phone) but the whole block of exception handling with e.pgcode/e.pgerror (and now e.diag) is surrounded by try statement that handles the case where e is **not** a psycopg2 Exception. That's fine.

I was wondering if, in that case, it would be smarter to put SQL statement output outside the psycopg2 specific block. What do you think ?

Another point : I saw that some parts of the `lib/messages.py` was only python2 compatible. I have a version which is python2/python3 compatible. Do you want me to push it ?

Reply:
> Alright it's crystal clear for me too now. I did not figure all the CI things out!
> So I changed params :
> 
> --debugging will have the purpose for developper to display backend pid and pause (and more maybe in the future ?)
> --details will display detailled errors at test failure

Still don't see the change at the PR, but you mean that debugging will be basically the same, but without the pause? If so, then yes. If not, having `--debugging` without the pause is what I'd like.

Reason: that way you can check at once, in one run at the CI job, if a problem at the same test(s) is the same for all OS and PostgreSQL versions or not. 

And not only useful for the CI, but even useful to debug locally as well, as you can use `ci-build` with several docker images to test and debug on your own :-)

> BTW my remark on mssql and the introduced bug is not relevant (I hadn't time to check the code with my phone) but the whole block of exception handling with e.pgcode/e.pgerror (and now e.diag) is surrounded by try statement that handles the case where e is not a psycopg2 Exception. That's fine.
> 
> I was wondering if, in that case, it would be smarter to put SQL statement output outside the psycopg2 specific block. What do you think ?

Makes sense. Right after:
```
print_error("Error running %s (%s)" % (test_desc, fname))
```

> Another point : I saw that some parts of the lib/messages.py was only python2 compatible. I have a version which is python2/python3 compatible. Do you want me to push it ?

Yeah, but better with a separate PR where we port that to python3 and verify that nothing else is broken or requires porting. We also need to verify that we still can work with python2 (CentOS6), so it's best if we don't mix that with this PR.

Reply:
Sorry I will be very busy (PGConf.eu time).
Two solutions : 
  * I can push all my changes for review (including python3.0 which is also python 2.7 compatible, I did some test on centos7 which is still python2.7 powered). 
  * We'll have to wait for a less stressful moment for me to make the changes

If one (or any) of you is on pgconf.eu let me know. We could meet. I have plans for using tds_fdw in an interesting context. I won't be difficult to reach me via twitter or any media.

If you're around don't hesitate!

Reply:
Sadly, I was not even aware of the conference. Otherwise maybe I could try having someone from my current team to explain how we use PostgreSQL at [Uyuni](https://www.uyuni-project.org/) its commercial relative [SUSE Manager ](https://www.suse.com/products/suse-manager/), the problems we face, and the solutions we took. Maybe next year! :-)

Let's do the following: push your changes and, if possible, give me [collaboration access](https://github.com/jcarnu/tds_fdw/settings/collaboration) (write option) at your forked repository. I will then make the small fixes to adapt the CI requirements, and will merge the changes.

Let me know if you see it as an acceptable plan.